### PR TITLE
Created xDnsServerForwarder resource

### PR DESCRIPTION
**Pull request by corydwood**
It uses the Get/Set-CimInstance cmdlets instead of the Get/Set-DnsServerForwarder cmdlets to allow compatibility with server OS's prior to Server 2012.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,6 +18,11 @@ Secondary zones allow client machine in primary DNS zones to do DNS resolution o
 * **xDnsServerZoneTransfer** This resource allows a DNS Server zone data to be replicated to another DNS server.
 * **xDnsRecord** This resource allwos for the creation of IPv4 host (A) records or CNames against a specific zone on the DNS server
 
+### xDnsServerForwarder
+
+* **IPAddress**: IP address of the forwarder
+* **RemoveAll**: Whether all forwarders should be removed or not
+* **Ensure**: Whether the forwarder should be present or removed
 
 ### xDnsServerPrimaryZone
```
Comment (TravisEz13) [inline README.md]: Please include a summary of changes in an unreleased version change per the [contribution guidelines](https://github.com/PowerShell/DscResources/blob/master/CONTRIBUTING.md#lifecycle-of-a-pull-request)

File: DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof [hunk 2]
```diff
@@ -0,0 +1,7 @@
+[ClassVersion("1.0.0"), FriendlyName("xDnsServerForwarder")]
+class MSFT_xDnsServerForwarder :  OMI_BaseResource
+{
+    [Key] String IPAddress;
+    [Write] Boolean RemoveAll;
```
Comment (TravisEz13) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof]: should this be ReplaceAll?    
Additionally there is an idempotency problem here.  If you want to specify multiple and replace make sure that the ones you specify, you will be removing the other ones you specified.  Causing your configuration to constantly churn.  If your IPAddress was an array, it would fix that, but then it cannot be the key.

If this is intended to only be specified once per configuration, that still might be ok, there is a pattern to follow, see [Guidance issue 80](https://github.com/PowerShell/DscResources/issues/80)

File: DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.psm1 [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    param
+    (
+        [Parameter(Mandatory)]
+        [string]$IPAddress,
+        [bool]$RemoveAll = $false,
+        [ValidateSet('Present','Absent')]
+        [string]$Ensure = 'Present'
+    )
+    Write-Verbose 'Getting DNS Forwarders'
+    Write-Output @{Forwarders = (Get-CimInstance -Namespace root\MicrosoftDNS -ClassName microsoftdns_server).Forwarders}
```
Comment (TravisEz13) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.psm1]: Your return from Get-TargetResource should match the schema you wrote in schema.mof

**Comment by TravisEz13** [comment]
It is required that you provide adequate test coverage for the code you change. [Contribution Guidelines -Adding test coverage for DSC resources](https://github.com/PowerShell/DscResources/blob/master/CONTRIBUTING.md#adding-test-coverage-for-dsc-resources)  Please work with us to come up with a plan to cover this code.

Suggestion: Base the tests on the [Unit tests Template](https://github.com/PowerShell/DscResources/blob/master/Tests.Template/unit_template.ps1)

File: DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof (hunk 2 above)
Comment (corydwood) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof]: @TravisEz13 It should be RemoveAll. If this property is $true, it ignores the IPAddress property because it will just remove all forwarders currently set on the DNS server if any exist.

From my experience, DSC isn't flexible enough to allow parameter sets for Key properties so that only one or the other is allowed. See the updated README.md for examples of each scenario.

**Comment by corydwood** [comment]
@TravisEz13 I do want to learn how to write Pester tests, but I haven't had the time to research that yet.

Most of the contributions I'm making are for resources I already have built and am using in production at my company (we have about 60). We've been given approval to share these on GitHub, so I was hoping I could contribute them without the tests and then I or someone in the community can add the tests later, since there's many in the Microsoft repositories already that don't have tests.

File: DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof (hunk 2 above)
Comment (PlagueHO) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof]: I agree with @TravisEz13 - it might be better to turn this into a SingleInstance resource with the IPAddress parameter as an array.

Something like this:
```
[ClassVersion("1.0.0"), FriendlyName("xDnsServerForwarder")]
class MSFT_xDnsServerForwarder :  OMI_BaseResource
{
    [Key, Description("Specifies the resource is a single instance, the value must be 'Yes'"), ValueMap{"Yes"}, Values{"Yes"}] String IsSingleInstance;
    [Write] String IPAddress[];
}
```

This would do away with the need for the "RemoveAll" parameter and may even simplify the code (assuming the CIM query code can take an array of IPAddresses) :smile:

Comment (corydwood) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof]: The reason I added the RemoveAll parameter was to allow you to remove all of the forwarders set on a DNS server without needing to know what they are. If I were to remove this parameter, you would have to know every IP Address that you would want to remove.

With the way forwarders are added to newly built DNS servers, you may not know what forwarders get added automatically. It could be the IPs set as the Primary or Secondary DNS servers on the network adapter. It could be the IPs currently set as forwarders on an existing DNS server in the domain.

Comment (PlagueHO) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof]: I understand why you did it, but I was just thinking If IPAddress was an array, to Remove all the forwarders you'd just set the array parameter to be an empty array (or not provide the parameter at all):
```powershell
xDnsServerForwarder RemoveAllForwardersfromDNSServer1 {
   IsSingleInstance = 'Yes'
   IPAddress = @()
}
```
That way the config never needs to know the current state of the resource (as you say) - it always replaces the current values if different.

Just my 2c :smile:

Comment (corydwood) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof]: I'd have to look into the SingleInstance option, since I haven't used that before, but from what I've seen, I believe the code I have currently is more flexible than using the SingleInstance method.

If it's configured as a single instance, it will replace all forwarders currently set on the server with the IPs in the IPAddress array. It wouldn't be able to just add/remove a single forwarder to/from an existing list of forwarders.

Comment (TravisEz13) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.schema.mof]: @PlagueHO 
Combining the single instance pattern and the membership pattern used in other resources would give us a schema like this.

This would allow the resource to be idempotent (specifically, prevent the configuration from churning on rerun), and allow the flexibility that @corydwood wants.  As well as using patterns DSC users should be familiar with.
```
xDnsServerForwarder RemoveAllForwardersfromDNSServer1 {
    [Key, Description("Specifies the resource is a single instance, the value must be 'Yes'"), ValueMap{"Yes"}, Values{"Yes"}] String IsSingleInstance;
    [Write, Description("IPAddresses of Forwarder in DNS Server which should match  exactly")] String IPAddress[];
    [Write, Description("Forwarder in DNS Server should include these IPAddresses")] String IPAddressToInclude[];
    [Write, Description("Forwarder in DNS Server should include these IPAddresses")] String IPAddressToExclude[];
}
```

**Comment by TravisEz13** [comment]
If others agree that it's worth it to take this without test coverage, you could file an issue to add test coverage.  On the other hand, the schema issues need to be address as changing it would be a breaking change.  So, getting the design of the schema correct needs to be resolved before we can merge the resource.

**Comment by PlagueHO** [comment]
I happy to contribute with test coverage if needed. Just raise an issue as @TravisEz13 says and I can add it in. But I'm definitely for getting the schema correct and preventing the resource from constantly churning. :+1:

**Comment by iainbrighton** [comment]
I'd go with the schema proposed by @PlagueHO. This matches the Group and xADGroup functionality whilst only requiring a single resource. Changing the schema after the first commit is... painful :confounded: 

@corydwood I'm happy to help out with tests too, but perhaps this is an ideal opportunity to start getting up-to-speed? I'd hate to start a precedent of allowing PRs without test coverage - I've found plenty of bugs (and still missed some) whilst writing them!

**Comment by TravisEz13** [comment]
I think I agree with @iainbrighton and @PlagueHO, because depending on implementation (assuming that the resource always replaced the complete list when it detects an inconsistency), it wouldn't be a breaking change to start with the schema @PlagueHO  proposed and then add the rest of the elements I later proposed.

**Comment by iainbrighton** [comment]
@corydwood I have moved the xADGroup "member" functionality into a common library in the xActiveDirectory module. I think you should just be able to grab the functions (and tests!) from here: https://github.com/iainbrighton/xActiveDirectory/blob/Issue27/DSCResources/MSFT_xADCommon/MSFT_xADCommon.ps1. You'll want the `Test-Members`, `Validate-MemberProperties` and `Remove-DuplicateMembers` functions.

In short, if Test-Members will return whether the `ExistingMembers` is compliant with the `Members`, `MembersToInclude` and `MembersToExclude` properties. It'll certainly get you a long way down the road :wink:

**Comment by corydwood** [comment]
@iainbrighton I'll look into those and update the code. Thanks.

**Comment by corydwood** [comment]
@TravisEz13, @PlagueHO, @iainbrighton: I've updated the code to make it a SingleInstance resource using the schema @PlagueHO recommended.

File: README.md [hunk 4]
```diff
@@ -65,6 +64,48 @@ Values include: { A-Record | C-Name }
...
+{
+    Import-DscResource -module xDnsServer
+    xDnsServerForwarder RemoveAllForwarders
+    {
+        IsSingleInstance = 'Yes'
+        IPAddresses = @()
+    }
+}
+Sample_Remove_All_Forwarders
+```
```
Comment (TravisEz13) [inline README.md]: Please include a summary of changes in an unreleased version change per the [contribution guidelines](https://github.com/PowerShell/DscResources/blob/master/CONTRIBUTING.md#lifecycle-of-a-pull-request)

**Comment by TravisEz13** [comment]
@PlagueHO @iainbrighton I did a quick review and it looks good to me.  I'll give you some time to comment.

File: DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.psm1 [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        [ValidateSet('Yes')]
+        [string]$IsSingleInstance,
+        [string[]]$IPAddresses
+    )
+    if (!$IPAddresses)
+    {
+        $IPAddresses = @()
+    }
+    Write-Verbose 'Setting DNS forwarders.'
+    Set-CimInstance -Namespace root\MicrosoftDNS -Query 'select * from microsoftdns_server' -Property @{Forwarders = $IPAddresses}
```
Comment (PlagueHO) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.psm1]: Any change you could break this line into multiple lines to make it 100 chars wide? E.g.
```powershell
Set-CimInstance `
-Namespace root\MicrosoftDNS `
-Query 'select * from microsoftdns_server' `
-Property @{Forwarders = $IPAddresses}
```

**Comment by PlagueHO** [comment]
The changes look pretty good to me. Would love to get unit tests into it, but as said I can help out with those fairly quickly. Should be able to put Integration tests in for this one as well. Although, as @iainbrighton mentioned we don't want to set a precedent here - would really like to avoid spending the rest of my life writing unit/integration tests :smiley:

File: DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.psm1 (hunk 5 above)
Comment (corydwood) [inline DSCResources/MSFT_xDnsServerForwarder/MSFT_xDnsServerForwarder.psm1]: @PlagueHO This is done.

**Comment by iainbrighton** [comment]
A quick scan looks good to me.

Are we planning on adding the IPAddressToInclude and IPAddressToExclude functionality in a future commit/version?

**Comment by DdenBraver** [comment]
@iainbrighton @corydwood Perhaps a "Ensure" parameter would help for future functionality.
If any help is needed with unit testing, I would be happy to help as well.

The following article helped me get up to speed with pester:
http://www.powershellmagazine.com/2014/03/12/get-started-with-pester-powershell-unit-testing-framework/

**Comment by TravisEz13** [comment]
@DdenBraver I'm not sure Ensure is the correct property for this.  You might need a read property for the Get method to indicate if it's in the desired state, but Since IPAddresses is an absolute list, you can add and remove without an Ensure property.

**Comment by DdenBraver** [comment]
@TravisEz13 That is correct, however I was thinking of the situation where you could use ensure = false. That time when you just want to enforce an IP address is not listed, while the rest can stay.

I did a quick test with this code today, and it worked flawlessly.

**Comment by TravisEz13** [comment]
@DdenBraver @iainbrighton  @corydwood I would suggest that this is an enhancement to the existing resource and we can accept the resource as is and design that at that time.  It should not affect the key.

**Comment by iainbrighton** [comment]
Agree. It won't be a breaking change.

File: Tests/Unit/MSFT_xDnsServerForwarder.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    #endregion
+}
+finally
+{
+    #region FOOTER
+    Restore-TestEnvironment -TestEnvironment $TestEnvironment
+    #endregion
+
+    # TODO: Other Optional Cleanup Code Goes Here...
+}
```
Comment (TravisEz13) [inline Tests/Unit/MSFT_xDnsServerForwarder.Tests.ps1]: I won't block the merge for this issue alone, but please be aware of this in the future (look at the PR for the symbol you should see here at the end of the diff.)
Files must end with a new line.  See [Style Guidelines, General Rule, #5](https://github.com/PowerShell/DscResources/blob/master/StyleGuidelines.md#general-rules)

**Comment by TravisEz13** [comment]
We have positive comments for contributor and it looks good to me.  I'm going to merge.  Thanks for your contribution.
